cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/TextToSpeechSSMLGenerator.scala

```diff
@@ -0,0 +1,113 @@
+    r.map(req => {
+      additionalHeaders.foreach(header => req.setHeader(header._1, header._2))
+      req
+    })
+  })
+
+  override protected def prepareEntity: Row => Option[AbstractHttpEntity] = { row =>
+    val localeValue = getValue(row, locale)
+    val zhCNVoiceName = "Microsoft Server Speech Text to Speech Voice (zh-CN, XiaomoNeural)"
+    val enUSVoiceName = "Microsoft Server Speech Text to Speech Voice (en-US, JennyNeural)"
+    val voiceName = if (localeValue == "zh-CN") zhCNVoiceName else enUSVoiceName
+    val textValue = getValue(row, text)
+    val body: String =
+      s"<speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis'" +
+        s" xmlns:mstts='https://www.w3.org/2001/mstts'" +
+        s" xml:lang='${localeValue}'><voice xml:lang='${localeValue}' xml:gender='Female' name='${voiceName}'>" +
+        s"<mstts:task name ='RoleStyle'/>${textValue}</voice></speak>"
+    Some(new StringEntity(body))
+  }
+
+  def formatSSML(row: Row, response: TextToSpeechSSMLResponse): String = {
+    val ssmlFormat: String = "<speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis' " +
+      "xmlns:mstts='https://www.w3.org/2001/mstts' xml:lang='%s'>".format(getValue(row, locale)) +
+      "%s</speak>"
+    val voiceFormat = "<voice name='%s'>".format(getValue(row, voice)) + "%s</voice>"
+    val expressAsFormat = "<mstts:express-as role='%s' style='%s'>%s</mstts:express-as>"
+    val builder = new StringBuilder()
+    val fullText = "%s".format(getValue(row, text))
+    var lastEnd = 0
+    response.Conversations.foreach(c => {
+      val content = c.Content
+      val role = c.Role.toLowerCase()
+      val style = c.Style.toLowerCase()
+      val begin = c.Begin
+      val end = c.End
+
+      val ssmlTurnRoleStyleStr = expressAsFormat.format(role, style, content)
+      val preStr = fullText.substring(lastEnd, begin - lastEnd)
+
+      if (preStr.length > 0) {
+        builder.append(preStr)
+      }
+      builder.append(ssmlTurnRoleStyleStr)
+      lastEnd = end
+    })
+
+    val endStr = fullText.substring(lastEnd)
+    if (endStr.length > 0) {
+      builder.append(endStr)
+    }
+    val outSsmlStr = ssmlFormat.format(voiceFormat.format(builder.toString())) + "\n"
+    outSsmlStr
+  }
+
+  val postprocessingTransformer: Transformer = new Transformer {
+    def transform(dataset: Dataset[_]): DataFrame = dataset.toDF().map { row =>
+      val response = row.getAs[String](getOutputCol).parseJson.convertTo[TextToSpeechSSMLResponse]
+      val result = formatSSML(row, response)
+      Row.fromSeq(row.toSeq ++ Seq(result))
+    }(RowEncoder(transformSchema(dataset.schema)))
+
+    override val uid: String = Identifiable.randomUID("TTSSSMLInternalPostProcessor")
+
+    override def copy(extra: ParamMap): Transformer = defaultCopy(extra)
+
+    override def transformSchema(schema: StructType): StructType = schema.add(getSSMLOutputCol, StringType)
+  }
```

Comment:
you can avoid the use of fancy dynamic typign here by using the UDFTransformer to apply a UDF to a column

Reply:
Addressed in most recent commit using a lambda transformer